python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py

```diff
@@ -0,0 +1,251 @@
+
+import yaml
+
+from semantic_kernel.connectors.ai.chat_completion_client_base import (
+    ChatCompletionClientBase,
+)
+from semantic_kernel.connectors.ai.open_ai.contents.open_ai_chat_message_content import OpenAIChatMessageContent
+from semantic_kernel.connectors.ai.open_ai.prompt_execution_settings.open_ai_prompt_execution_settings import (
+    OpenAIChatPromptExecutionSettings,
+)
+from semantic_kernel.connectors.ai.open_ai.utils import (
+    get_tool_call_object,
+)
+from semantic_kernel.contents.chat_history import ChatHistory
+from semantic_kernel.exceptions.planner_exceptions import PlannerInvalidConfigurationError
+from semantic_kernel.functions.kernel_arguments import KernelArguments
+from semantic_kernel.functions.kernel_function import KernelFunction
+from semantic_kernel.kernel import Kernel
+from semantic_kernel.kernel_pydantic import KernelBaseModel
+from semantic_kernel.planners.function_calling_stepwise_planner.function_calling_stepwise_planner_options import (
+    FunctionCallingStepwisePlannerOptions,
+)
+from semantic_kernel.planners.function_calling_stepwise_planner.function_calling_stepwise_planner_result import (
+    FunctionCallingStepwisePlannerResult,
+    UserInteraction,
+)
+from semantic_kernel.planners.planner_extensions import PlannerKernelExtension
+from semantic_kernel.prompt_template.kernel_prompt_template import KernelPromptTemplate
+from semantic_kernel.prompt_template.prompt_template_config import PromptTemplateConfig
+
+CUR_DIR = os.path.dirname(os.path.realpath(__file__))
+PLAN_YAML_FILE_PATH = os.path.join(CUR_DIR, "generate_plan.yaml")
+STEP_PROMPT_FILE_PATH = os.path.join(CUR_DIR, "step_prompt.txt")
+
+STEPWISE_PLANNER_PLUGIN_NAME = "StepwisePlanner_Excluded"
+
+STEPWISE_USER_MESSAGE = (
+    "Perform the next step of the plan if there is more work to do."
+    "When you have reached a final answer, use the UserInteraction-SendFinalAnswer"
+    "function to communicate this back to the user."
+)
+
+USER_INTERACTION_SEND_FINAL_ANSWER = "UserInteraction-SendFinalAnswer"
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class FunctionCallingStepwisePlanner(KernelBaseModel):
+    service_id: str
+    options: FunctionCallingStepwisePlannerOptions
+    generate_plan_yaml: str
+    step_prompt: str
+
+    def __init__(self, service_id: str, options: Optional[FunctionCallingStepwisePlannerOptions] = None):
+        """Initialize a new instance of the FunctionCallingStepwisePlanner
+
+        Args:
+            service_id (str): The service id
+            options (Optional[FunctionCallingStepwisePlannerOptions], optional): The options for
+                the function calling stepwise planner. Defaults to None.
+        """
+        options = options or FunctionCallingStepwisePlannerOptions()
+        generate_plan_yaml = (
+            options.get_initial_plan()
+            if options.get_initial_plan
+            else (lambda path: open(path).read())(PLAN_YAML_FILE_PATH)
```

Comment:
Sorry, do you need this to be a lambda? Doesn't `else open(PLAN_YAML_FILE_PATH).read()` works?

Reply:
Just a stylistic choice here, TBH. My preference would be to use a context manager, like I had previously with the read_file function, so that I know the file is closed properly after reading. Then to use a lambda, it may be a little overkill, but was thinking in terms of resource management again. I can simplify it, if that doesn't jive with you. :)

Reply:
Oh! I see, in that case let's have a helper function here in this module but more targeted to this logic and the one on line 79. Something like this:
```python
def call_attr(object: Any, attr_name: str, fallback_file_path: str) -> str:
    attr_value = getattr(obj, attr_name, None)
    if attr_value:
        return attr_value()

    with open(fallback_file_path) as fallback_file:
        return fallback_file.read()
```

Then use `call_attr` in both places.
What do you think?